server/data/data-library/src/main/java/org/apache/james/user/lib/AbstractUsersRepository.java

```diff
@@ -123,6 +124,13 @@ public boolean supportVirtualHosting() {
 
     @Override
     public Username getUser(MailAddress mailAddress) throws UsersRepositoryException {
+        Username virtualHostingNormalizedUsername = applyVirtualHosting(mailAddress);
+        return Optional.ofNullable(getUserByName(virtualHostingNormalizedUsername))
+            .map(User::getUserName)
```

Comment:
you are now doing an explicit backend request when you call getUser(). Before it was just a string substitution. Did you measure the potential impact of this change?

Reply:
> you are now doing an explicit backend request when you call getUser(). Before it was just a string substitution. Did you measure the potential impact of this change?

Yes. 

Is this a *performance* versus *correctness* debate ? I remember what's the team answer had been every time I propose to trade correctness to get performance...

Tracking the calls:
 - There's some webadmin routes (groups, alias, forward) -> not critical
 - There's some SMTP Hooks RCPT checks & Sender checks -> not chocking
 - ~ 10 mailets -> I don't think it is critical.

Also some calls are directly followed by a `contains(Username)`. `containsUserWithMailAddress(MailAddress)` would allow us to further optimize that.

Furthermore it is metadata,  username reading had been a bottleneck in none of the performance tests I have seen so far.

Reply:
> Is this a performance versus correctness debate ?

No, it's just a "did you think to mesure performance changed introduced by this PR?".

I'm a little worried by the ~10 mailets using it though. This means that if you are using 5 of them, you are doing 5 more requests per mail you receive.

Reply:
>  5 more requests per mail you receive

5 more requests on tiny, cached metadata.

I could plan a performance test on mail processing if it makes you feel better (but the impact is going to be veeeeery limited, if not measurable at all)